helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizer.java

```diff
@@ -0,0 +1,347 @@
+    clusterState.instances = dataAccessor.getChildNames(dataAccessor.keyBuilder().liveInstances());
+    clusterState.instances.addAll(inputFields.newInstances);
+
+    // Check if to be removed instances and old instances in swap node exist in live instance
+    if (!inputFields.nodeSwap.isEmpty() || !inputFields.instancesToRemove.isEmpty()) {
+      Set<String> liveInstanceSet = new HashSet<>(clusterState.instances);
+      for (Map.Entry<String, String> nodeSwapPair : inputFields.nodeSwap.entrySet()) {
+        if (!liveInstanceSet.contains(nodeSwapPair.getKey())) {
+          throw new InvalidParameterException("Invalid input: instance [" + nodeSwapPair.getKey()
+              + "] in SwapInstances does not exist in cluster.");
+        }
+      }
+      for (String instanceToRemove : inputFields.instancesToRemove) {
+        if (!liveInstanceSet.contains(instanceToRemove)) {
+          throw new InvalidParameterException("Invalid input: instance [" + instanceToRemove
+              + "] in RemoveInstances does not exist in cluster.");
+        }
+      }
+      if (!inputFields.instancesToRemove.isEmpty()) {
+        clusterState.instances.removeIf(inputFields.instancesToRemove::contains);
+      }
+    }
+
+    // Read instance and cluster config.
+    // It will throw exception if there is no instanceConfig for newly added instance.
+    for (String instance : clusterState.instances) {
+      InstanceConfig config = cfgAccessor.getInstanceConfig(clusterId, instance);
+      clusterState.instanceConfigs.add(config);
+    }
+    clusterState.clusterConfig = cfgAccessor.getClusterConfig(clusterId);
+    return clusterState;
+  }
+
+  private AssignmentResult computeOptimalAssignmentForResources(InputFields inputFields,
+      ClusterState clusterState, String clusterId) throws Exception {
+
+    AssignmentResult result = new AssignmentResult();
+    // Iterate through resources, read resource level info and get potential assignment.
+    HelixDataAccessor dataAccessor = getDataAccssor(clusterId);
+    List<IdealState> wagedResourceIdealState = new ArrayList<>();
+
+    for (String resource : clusterState.resources) {
+      IdealState idealState =
+          dataAccessor.getProperty(dataAccessor.keyBuilder().idealStates(resource));
+      // Compute all Waged resources in a batch later.
+      if (idealState.getRebalancerClassName() != null && idealState.getRebalancerClassName()
+          .equals(WagedRebalancer.class.getName())) {
+        wagedResourceIdealState.add(idealState);
+        continue;
+      }
+      // For non Waged resources, we don't compute resources not in white list.
+      if (!inputFields.resourceFilter.isEmpty() && !inputFields.resourceFilter.contains(resource)) {
+        continue;
+      }
+      // Use getIdealAssignmentForFullAuto for FULL_AUTO resource.
+      if (idealState.getRebalanceMode() == IdealState.RebalanceMode.FULL_AUTO) {
```

Comment:
It's worth to have private functions to implement FULL_AUTO and SEMI_AUTO scenarios. Then this compute function would be much clear and organized.

Reply:
Thanks for suggestion. I have tried and actually I am not sure if adding a separate function for several lines with many parameters would be more organized...